gtsfm/utils/splat.py

```diff
@@ -312,48 +313,139 @@ def transform_gaussian(gaussianA: dict, bSa: gtsam.Similarity3) -> dict:
+
+    if means.ndim == 2:
+        converted_means = torch.matmul(means, R.T) + t
+    else:
+        converted_means = torch.matmul(means, R.T) + t.unsqueeze(0)
+    converted_means = converted_means * s
+
+    sim3_quat_coeffs = torch.tensor(bSa.rotation().toQuaternion().coeffs(), dtype=dtype, device=device)
+    sim3_quat_wxyz = sim3_quat_coeffs[[3, 0, 1, 2]]
+
+    if is_dict_input:
+        quats_wxyz = gaussian_splats["quats"]
+        converted_quaternions = _apply_quaternion_left_multiply(
+            sim3_quat_wxyz.to(dtype=quats_wxyz.dtype, device=quats_wxyz.device),
+            quats_wxyz,
+        )
+        scales_tensor = gaussian_splats["scales"]
+        scale_offset = torch.log(torch.tensor(scale_value, dtype=scales_tensor.dtype, device=scales_tensor.device))
+        converted_scales = scales_tensor + scale_offset
+
+        converted_gaussian_splats = gaussian_splats.copy()
+        converted_gaussian_splats["means"] = converted_means
+        converted_gaussian_splats["quats"] = converted_quaternions
+        converted_gaussian_splats["scales"] = converted_scales
+        return converted_gaussian_splats
+
+    rotations_xyzw = gaussian_splats.rotations
+    rotations_wxyz = torch.cat([rotations_xyzw[..., 3:], rotations_xyzw[..., :3]], dim=-1)
+    converted_rotations_wxyz = _apply_quaternion_left_multiply(
+        sim3_quat_wxyz.to(dtype=rotations_xyzw.dtype, device=rotations_xyzw.device),
+        rotations_wxyz,
+    )
+    converted_rotations_xyzw = torch.cat([converted_rotations_wxyz[..., 1:], converted_rotations_wxyz[..., :1]], dim=-1)
 
-    converted_means = s * ((R @ means.T).T + t)
+    converted_scales = gaussian_splats.scales * torch.tensor(
+        scale_value, dtype=gaussian_splats.scales.dtype, device=gaussian_splats.scales.device
+    )
 
-    w0, x0, y0, z0 = bSa.rotation().toQuaternion().coeffs()[[3, 0, 1, 2]]
-    w1, x1, y1, z1 = gaussian_splats["quats"].unbind(-1)
+    if not is_dataclass(gaussian_splats):
+        raise TypeError("Expected gaussian_splats to be a dataclass implementing GaussiansProtocol.")
 
-    converted_quaternions_w = w0 * w1 - x0 * x1 - y0 * y1 - z0 * z1
-    converted_quaternions_x = w0 * x1 + x0 * w1 + y0 * z1 - z0 * y1
-    converted_quaternions_y = w0 * y1 - x0 * z1 + y0 * w1 + z0 * x1
-    converted_quaternions_z = w0 * z1 + x0 * y1 - y0 * x1 + z0 * w1
+    replace_kwargs = {
+        "means": converted_means,
+        "scales": converted_scales,
+        "rotations": converted_rotations_xyzw,
+    }
+    converted_gaussian_splats = replace(gaussian_splats, **replace_kwargs)
+    return converted_gaussian_splats
 
-    converted_quaternions = torch.stack(
-        [converted_quaternions_w, converted_quaternions_x, converted_quaternions_y, converted_quaternions_z], dim=1
-    )
 
-    converted_quaternions = converted_quaternions / converted_quaternions.norm(dim=-1, keepdim=True)
+def merge_gaussian_splats(
+    gaussians_a: dict[str, torch.Tensor] | GaussiansProtocol,
```

Comment:
Why all the hedging? Do we have multiple representations? I guess maybe AnySplat and gsplat ?